sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala

```diff
@@ -202,11 +204,15 @@ private[parquet] class ParquetRowConverter(
 
   override def start(): Unit = {
     var i = 0
-    while (i < currentRow.numFields) {
+    while (i < fieldConverters.length) {
       fieldConverters(i).updater.start()
       currentRow.setNullAt(i)
```

Comment:
Yea, I think it should be fine to do `setNullAt` at non-corresponding field, right?

Reply:
Yes. The following while loop at https://github.com/apache/spark/blob/6b19f579e5424b5a8c16d6817c5a59b9828efec2/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala#L212-L215 ensures all remaining columns/fields in the current row are nulled-out.

Reply:
I am also lost here. The `i` index seems to me following the parquet fields so is not the `updater.ordinal` the the correct index to update the currentRow?

I would expect something like:

`scala
val updater = fieldConverters(i).updater
updater.start()
currentRow.setNull(updater.ordinal)
`

Reply:
I can see how this is confusing. As part of the `start` method, all columns in the current row must be set to null. Some of those columns are set to null in

https://github.com/apache/spark/blob/6b19f579e5424b5a8c16d6817c5a59b9828efec2/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala#L207-L211

The rest of them are set to null in

https://github.com/apache/spark/blob/6b19f579e5424b5a8c16d6817c5a59b9828efec2/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala#L212-L215

This is equivalent to

```scala
var i = 0
while (i < fieldConverters.length) {
  fieldConverters(i).updater.start()
  i += 1
}
var j = 0
while (j < currentRow.numFields) {
  currentRow.setNullAt(j)
  j += 1
}
```

Is that clearer? Maybe I should rewrite the code that way.

Reply:
Ehhh... the links in the previous post didn't come through as code blocks... Don't know why... Let me try to fix that...

Reply:
@viirya @attilapiros Hi guys. Does my explanation make sense? If so, do you want me to change the code as I suggested or leave it as-is in the current PR commit?

Reply:
I see. I think doing this separately is better and you can rewrite it to one-liner, like for setNullAt: 

```scala
(0 until currentRow.numFields).foreach(currentRow.setNullAt)
```

Reply:
Thank you both for your feedback.

> Seems It can save some redundant iterations.

That was my motivation in writing the code this way. While the code is not as clear as it could be, it is very performance critical.

I'm going to push a new commit keeping the current code but with a brief explanatory comment.

Reply:
> I'm going to push a new commit keeping the current code but with a brief explanatory comment.

On further careful consideration, I believe that separating the calls to `currentRow.setNullAt(i)` into their own loop actually won't incur any significant performance degradation—if any at all.

The performance of the `start()` method is dominated by the calls to `fieldConverters(i).updater.start()` and `currentRow.setNullAt(i)`. Putting the latter calls into their own loop won't change the count of those method calls, just the order. @viirya LMK if you disagree with my analysis.

I will push a new commit with separate while loops. I won't use the more elegant `(0 until currentRow.numFields).foreach(currentRow.setNullAt)` because that's not a loop, and I doubt either the Spark or Hotspot optimizer can turn that into a loop.